src/hotspot/share/utilities/debug.cpp

```diff
@@ -67,6 +67,11 @@
 #include <stdio.h>
 #include <stdarg.h>
 
+#if defined(TARGET_COMPILER_gcc)
+#undef JNIEXPORT
```

Comment:
This partially reverts [JDK-8264593](https://bugs.openjdk.org/browse/JDK-8264593); @kevinjwalls will this work for you?

Reply:
No, it doesn't. Even after [JDK-8264593](https://bugs.openjdk.org/browse/JDK-8264593) the symbols from debug.cpp were not exported on Linux, since they were not mentioned in `make/data/hotspot-symbols`, and thus were made local.

The reason I put this in was to keep parity with the symbol table before and after the patch.

My guess is that it was only on Windows it was necessary that the functions where JNIEXPORTed, otherwise someone would have complained after JDK-8264593. But for platform consistency, it might make sense to export these even on gcc as well. After this PR is integrated, this is easy to do -- just remove the above hack.

Reply:
they were absent from .dynsym but available in .symtab as bind: local, visibility: default. Now they are completely gone.

The symbols were supposed to be used from gdb. I'm not sure if gdb can access local symbols, so it's possible that they didn't work even before this change.

Reply:
> But for platform consistency, it might make sense to export these even on gcc as well.
JDK\-8288293 would definitely like that a lot

Reply:
Thanks - when I last tried, e.g. "call universe()" would work on Linux/gdb, so I think gdb did resolve a local symbol (which cannot be found by callling dlsym, but can be found in the binary symboltable).
On Windows the debug.cpp symbols just weren't in the binary at all, then with JDK-8264593 JNIEXPORT they were, and appeared in the .map file we generate also.

So I am saying I added the JNIEXPORT for Windows, to get the debug.cpp symbols into the binary.
This removes JNIEXPORT for gcc, which already had the debug.cpp symbols in the binary.

But I will go check this...

Reply:
> they were absent from .dynsym but available in .symtab as bind: local, visibility: default. Now they are completely gone.

I don't think that statement is correct. Or else my understanding about symbol tables is incomplete. Both `nm` (which shows `.symtab`) and `nm -D` (which shows `.dynsym`) are identical for the libjvm.so before and after the patch wrt to the JNIEXPORTed debug.cpp functions.

Was this a statement of your understanding of the situation, or could you spot an actual difference? If it is the latter, please let me know how I can view this difference.

Reply:
On a latest JDK build today, in Linux I can attach to a live process (not a core dump):
  gdb --pid PID
..and type:
  call universe()
(I think you need to switch to a JavaThread first) and see on the JVM's output:

"Executing universe"
Heap
 garbage-first heap   total reserved 30932992K,  ...etc...
 

Removing JNIEXPORT here on Linux and rebuilding: still works.

readelf shows: 
existing build (using JNIEXPORT):
 32252: 0000000000698160   249 FUNC    LOCAL  DEFAULT   11 universe

without JNIEXPORT:
 32252: 0000000000698160   249 FUNC    LOCAL  HIDDEN    11 universe

So gdb on Linux can call a HIDDEN function, i.e. without the JNIEXPORT.  It seems happy either way, not affected by visibility.

Reply:
Windows:
Attach Visual Studio,
in the Immediate Window, type (funny VS syntax I do not claim to understand):

{,,JVM}universe()

...and on the JVM's console window I see:

"Executing universe"
Heap
 garbage-first heap   total reserved 8343552K,...etc...

(It actually did nothing on the first invocation for me, I have typed it 5 times, only got 4 sets of output.  But the symbol is working...)

In that working (JNIEXPORT) case, the map file jdk/bin/server/jvm.dll.map shows:

 0001:0027ff50       universe                   0000000180280f50 f   debug.obj

(I do like that we call it jvm.dll.map now, not just "jvm.map"!)

Test removing JNIEXPORT:
 
Without JNIEXPORT, on Windows, the symbol universe is just not there.  Nothing in the map file.  In Visual Studio:

{,,JVM}universe()
identifier "universe" is undefined

That confirms the expected Windows behaviour: we still need JNIEXPORT to make debug.cpp utils callable by a debugger.

Reply:
> (I do like that we call it jvm.dll.map now, not just "jvm.map"!)

You can thank @fthevenet for that! :)

Reply:
> Removing JNIEXPORT here on Linux and rebuilding: still works.
> 
> readelf shows: existing build (using JNIEXPORT): 32252: 0000000000698160 249 FUNC LOCAL DEFAULT 11 universe
> 
> without JNIEXPORT: 32252: 0000000000698160 249 FUNC LOCAL HIDDEN 11 universe
> 
> So gdb on Linux can call a HIDDEN function, i.e. without the JNIEXPORT. It seems happy either way, not affected by visibility.

@kevinjwalls Thank you for testing this for me!